aten/src/ATen/core/type.cpp

```diff
@@ -1126,35 +1124,53 @@ void ClassType::checkNotExist(const std::string& name, const std::string& what)
   }
 
   // Check no overlap with existing attributes
-  for (size_t i = 0; i < attributeNames_.size(); ++i) {
+  for (size_t i = 0; i < attributes_.size(); ++i) {
     TORCH_CHECK(
-        name != attributeNames_[i],
+        name != attributes_[i].getName(),
         "attempting to add ",
         what,
         " '",
         name,
         "' to ",
         python_str(),
         " but an attribute field of the same name already exists with type ",
-        attributeTypes_[i]->python_str());
+        attributes_[i].getType()->python_str());
   }
 }
 
+void ClassType::provideNewAttributeKind(AttributeKind attributeKind) {
+    attributes_.push_back(attributeKind);
+    attributeTypes_.push_back(attributeKind.getType());
+    AT_ASSERT(attributes_.size() == attributeTypes_.size());
+}
+
 size_t ClassType::addAttribute(
     const std::string& name,
     const TypePtr& type,
     bool is_parameter,
-    bool allow_any) {
-  const char* what = is_parameter ? "parameter" : "attribute";
-  checkNotExist(name, what);
+    bool allow_any,
+    bool was_registered_as_buffer) {
 
+  std::string what = is_parameter ? "parameter" : "attribute";
+  what += (was_registered_as_buffer? "buffer" : "not buffer");
+  checkNotExist(name, what);
   if (!allow_any) {
-    checkNoAny(*this, what, name, type);
+    checkNoAny(*this, what.c_str(), name, type);
+  }
+
+  size_t slot = attributes_.size();
+
+  ATTRIBUTE_KIND_ENUM kind = ATTRIBUTE_KIND_ENUM::REGULAR_ATTRIBUTE;
```

Comment:
Should we consider just making `ATTRIBUTE_KIND_ENUM` something that the user sets? Hiding allows some weird behavior. For example: right now you can call `addAttribute` and have something be both a parameter and buffer—this code would silently mark the attribute as a parameter.

Reply:
I'd rather keep it as a detail of our system, and then provide safeguards. Users already signal if something is a buffer or parameter by how they register it with our module.

Reply:
curious what happens in eager if someone registers the same name/value pair as a parameter and buffer…

Reply:
I'll add a test for this!